extensions/auth/opa/impl/src/main/java/org/apache/polaris/extension/auth/opa/token/FileBearerTokenProvider.java

```diff
@@ -0,0 +1,307 @@
+      if (refreshTime.isBefore(minRefreshTime)) {
+        logger.warn(
+            "JWT expires too soon ({}), using minimum refresh interval instead", expiration.get());
+        return lastRefresh.plus(refreshInterval);
+      }
+
+      logger.debug(
+          "Using JWT expiration-based refresh: token expires at {}, refreshing at {}",
+          expiration.get(),
+          refreshTime);
+      return refreshTime;
+    }
+
+    // Fall back to fixed interval (token is not a valid JWT or has no expiration)
+    logger.debug("Token is not a valid JWT or has no expiration, using fixed refresh interval");
+    return lastRefresh.plus(refreshInterval);
+  }
+
+  @Nullable
+  private String loadTokenFromFile() {
+    int attempts = 0;
+    long deadlineMs =
+        clock.millis()
+            + 3000; // 3 second deadline for retries (reasonable for CSI driver file mounts)
+    String currentCachedToken = cachedToken; // Snapshot of current cache
+
+    while (true) {
+      try {
+        // Check if file is readable first
+        if (!Files.isReadable(tokenFilePath)) {
+          // Treat as transient error and retry (could be CSI driver not ready yet)
+          throw new IOException("File is not readable: " + tokenFilePath);
+        }
+
+        String token = Files.readString(tokenFilePath, StandardCharsets.UTF_8).trim();
+        if (!token.isEmpty()) {
+          return token;
+        }
+
+        // Empty token - treat as transient issue (could be file being written)
+        throw new IOException("File contains only whitespace: " + tokenFilePath);
+      } catch (IOException e) {
+        // Treat file system exceptions as transient (file being rotated, temporary unavailability)
+        logger.debug(
+            "Transient error reading token file (attempt {}): {}", attempts + 1, e.getMessage());
+      }
+
+      // If we have a cached token and it's safe to use, fall back to it
+      if (currentCachedToken != null && !currentCachedToken.isEmpty()) {
+        logger.warn(
+            "Failed to read new token from file after {} attempts, using cached token",
+            attempts + 1);
+        return currentCachedToken;
+      }
+
+      // No cached token available and we can't read from file
+      attempts++;
+      if (attempts >= 5 || clock.millis() > deadlineMs) {
+        // Return null to let refreshToken() decide how to handle this based on cached token state
+        logger.debug("Token unavailable after {} attempts", attempts);
+        return null;
+      }
+
+      // Exponential backoff: 100ms, 200ms, 400ms, 800ms, 1000ms (reasonable for CSI driver
+      // scenarios)
+      long delayMs = Math.min(100L << (attempts - 1), 1000);
+      try {
+        Thread.sleep(delayMs);
+      } catch (InterruptedException ie) {
+        Thread.currentThread().interrupt();
+        logger.debug("Interrupted while reading token, returning null");
+        return null;
+      }
+    }
+  }
```

Comment:
I think, the whole function can be simplified like this:
```suggestion
  private String loadTokenFromFile() {
    try {
      String token = Files.readString(tokenFilePath, StandardCharsets.UTF_8).trim();
      if (!token.isEmpty()) {
        return token;
      }
    } catch (IOException e) {
      logger.debug("Failed to read token from file", e);
    }
    return null;
  }
```